Add support for creating certificates from CSRs

djc opened the pull request:
Depends on https://github.com/rusticata/x509-parser/pull/58 and (transitively) https://github.com/rusticata/oid-registry/pull/1, but if you have any early feedback that would be great!

djc commented:
The upstream PRs have been merged (but not yet released).

est31 commented:
Thanks for this PR! I'll review it more closely later but as a remark for the start, I'm not sure if the `from_name` function and the TryFrom impl should be added. This makes the public rcgen API use types from the `x509-parser` crate which means that semver breaking updates cause us to have them too. In general though, it's good to have CSR support in rcgen.

est31 commented:
Also: could you add tests? :)

djc commented:
About the semver compatibility issues:

- `DistinguishedName::from_name()` is actually private, so I don't think that should be a concern?
- The main entry point `CertificateSigningRequest::from_x509()` itself exposes an x509-parser type in the public API. If it would take DER or PEM bytes, that means either (a) rcgen should do signature verification or (b) the CSR has to be parsed twice if the users wants to also do signature verification. What do you prefer, or do you see another option?

Will add some tests.

est31 commented:
> from_name() is actually private

good point, case rested.

> either (a) rcgen should do signature verification or (b) the CSR has to be parsed twice if the users wants to also do signature verification. What do you prefer, or do you see another option?

Is it hard to do (a)? rcgen's main task is to create certificates, where you can't do much wrong, but doing signature verification wrong doesn't seem *that* risky, as an attacker having a certificate for a public key they don't own doesn't give them *that* many more capabilities. But the same time, it's not the goal of rcgen to verify things, but to create them.

djc commented:
I changed it to verify the signature using the signature verification feature from x509-parser (which I wrote for their CSR support -- this seemed more convenient than reproducing the verification logic in rcgen. I've also moved the `TryFrom` impl to a private `try_from()` method on `SanType`. Let me know if this is okay, then I'll write up some tests next.

est31 commented:
@djc sounds like a plan! :+1:

djc commented:
Updated to use x509-parser 0.9.

est31 commented:
Wonderful! Will take a look at it later today.

est31 commented on src/lib.rs:
```diff
@@ -156,6 +156,19 @@ pub enum SanType {
 
 impl SanType {
+	#[cfg(feature = "x509-parser")]
+	fn try_from(name: &x509_parser::extensions::GeneralName<'_>) -> Result<Self, RcgenError> {
```
Could you rename this to `try_from_general` or something? `try_from` is very generic, makes mainly sense if part of the TryFrom trait impl. Also: colon placement.

est31 commented on src/lib.rs:
```diff
@@ -403,6 +416,31 @@ impl DistinguishedName {
 	}
+
+	#[cfg(feature = "x509-parser")]
+	fn from_name(name: &x509_parser::x509::X509Name) -> Result<Self, RcgenError> {
```
colon placement

est31 commented on src/lib.rs:
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
+}
+
+impl CertificateSigningRequest {
+	/// Parse certificate signing request from the ASCII PEM format
```
```suggestion
	/// Parse a certificate signing request from the ASCII PEM format
```

est31 commented on src/lib.rs:
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
+		Self::from_der(&csr.contents)
+	}
+
+	/// Parse certificate signing request from DER-encoded bytes
```
```suggestion
	/// Parse a certificate signing request from DER-encoded bytes
```

est31 commented on src/lib.rs:
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
+
+	/// Parse certificate signing request from DER-encoded bytes
+	///
+	/// Currently, this only supports the `Subject Alternative Name` extension.
```
Maybe say here that for other extensions, an error is returned.

est31 commented on src/lib.rs:
```diff
@@ -424,6 +462,101 @@ impl <'a> Iterator for DistinguishedNameIterator<'a> {
+	}
+	/// Serializes the requested certificate, signed with another certificate's key, to the ASCII PEM format
+	///
+	/// *This function is only available if rcgen is built with the "pem" feature*
```
Thanks for adding this! :heart:

est31 reviewed (changes requested):
Ok except for some nits.

djc commented:
Addressed all the nits.

est31 reviewed (approved):
Thanks! It's a bit sad that the nom update brought bit-vec into our lockfile which has way too many dependencies for the simple task it's doing, but that's more of a long term problem. Merging.

djc commented:
Yeah, x509-parser should maybe default to depending on nom with default-features = false... Thanks for merging! Do you plan on publishing a new release soon?

est31 commented:
@djc waiting for travis CI to pass, then I'll release. After that I'll switch to github actions because I don't have patience :).

djc commented:
Yeah, I was surprised you don't have Actions here, yet!

est31 commented:
0.8.6 is out!

djc commented:
Thanks!
